File: bisection.cpp
#include <iostream>
#include <cmath>

using namespace std;

// Define the quadratic equation: x^3 - 4x - 9
float equation(float x)
{
    return (x * x * x) - (4 * x) - 9;
}

// Bisection method function
float bisection(float func(float), float a, float b, float tolerance, int maxIterations)
{
    for (int i = 1; i <= maxIterations; i++)
    {
        float root = (a + b) / 2.0;
        float fx = func(root);

        cout << i << "\t\t" << root << "\t" << fx << endl;

        if (fabs(fx) < tolerance)
        {
            cout << "\nRoot found: " << root << endl;
            return root; // Return the root when found within tolerance
        }

        if (fx * func(a) < 0)
        {
            b = root; // Update the interval
        }
        else
        {
            a = root; // Update the interval
        }
    }

    cout << "\nRoot not found within " << maxIterations << " iterations." << endl;
    return NAN; // Return NaN if root not found
}

int main()
{
    float a = 1.0;           // Left endpoint of the interval
    float b = 3.0;           // Right endpoint of the interval
    float tolerance = 1e-6;  // Tolerance for stopping criterion
    int maxIterations = 100; // Maximum number of iterations

    cout << "Iteration\tRoot\tf(x)" << endl;

    bisection(equation, a, b, tolerance, maxIterations);

    return 0;
}

File: bisection.py
# Define the quadratic equation: x^3 - 4x - 9
def quadratic_equation(x):
    return x**3 - 4*x - 9

# Bisection method function
def bisection_method(func, interval_start, interval_end, tolerance, max_iterations):
    for iteration in range(1, max_iterations + 1):
        root = (interval_start + interval_end) / 2.0
        function_value = func(root)

        print(f"{iteration}\t\t{root:.6f}\t{function_value:.6f}")

        if abs(function_value) < tolerance:
            print(f"\nRoot found: {root:.6f}")
            return root  # Return the root when found within tolerance

        if function_value * func(interval_start) < 0:
            interval_end = root  # Update the interval
        else:
            interval_start = root  # Update the interval

    print(f"\nRoot not found within {max_iterations} iterations.")
    return None  # Return None if root not found

# Main function
def main():
    interval_start = 1.0  # Left endpoint of the interval
    interval_end = 3.0  # Right endpoint of the interval
    tolerance = 1e-6  # Tolerance for stopping criterion
    max_iterations = 100  # Maximum number of iterations

    print("Iteration\tRoot\tf(x)")

    bisection_method(quadratic_equation, interval_start, interval_end, tolerance, max_iterations)

if __name__ == "__main__":
    main()
#  if you would like to avoid using if __name__ == "__main__":, you can directly call the main() function without the need for this construct.